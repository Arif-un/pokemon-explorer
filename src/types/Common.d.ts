export interface QueryParams {
  types?: string
  sort?: string
  limit?: number
  offset?: number
  name?: string
}
