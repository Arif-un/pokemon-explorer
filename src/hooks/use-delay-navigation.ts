import { useState } from 'react'

import { useRouter } from '@tanstack/react-router'

interface NextLocation {
  to: string
  replace: boolean
}

export function useDelayedNavigation() {
  const router = useRouter()
  const [isNavigating, setIsNavigating] = useState(false)
  const [nextLocation, setNextLocation] = useState<NextLocation | null>(null)

  const navigate = (to: string, { replace = false } = {}) => {
    setIsNavigating(true)
    setNextLocation({ to, replace })
  }

  const finishNavigation = () => {
    if (nextLocation) {
      const { to, replace } = nextLocation
      router.navigate({ to, replace, search: prv => ({ ...prv }) })
      setIsNavigating(false)
      setNextLocation(null)
    }
  }

  return { isNavigating, navigate, finishNavigation }
}
