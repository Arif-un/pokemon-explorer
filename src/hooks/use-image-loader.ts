import { useEffect, useState } from 'react'

interface ImageState {
  src: string
  isLoading: boolean
  error: Event | string | null
}

const imageCache: Record<string, boolean> = {}

export function useImageLoader(src: string, fallbackSrc: string): ImageState {
  const [imageState, setImageState] = useState<ImageState>({
    src: fallbackSrc,
    isLoading: true,
    error: null
  })

  useEffect(() => {
    if (imageCache[src]) {
      setImageState({
        src: src,
        isLoading: false,
        error: null
      })
      return
    }

    const img = new Image()
    img.src = src

    img.onload = () => {
      imageCache[src] = true

      setImageState({
        src: src,
        isLoading: false,
        error: null
      })
    }

    img.onerror = error => {
      setImageState({
        src: fallbackSrc,
        isLoading: false,
        error
      })
    }

    return () => {
      img.onload = null
      img.onerror = null
    }
  }, [src, fallbackSrc])

  return imageState
}
