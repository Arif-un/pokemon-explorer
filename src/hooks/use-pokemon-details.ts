import { useQuery } from '@tanstack/react-query'

import { fetchPokemonDetails } from '@/services/pokemon-api'
import { type PokemonDetailsResponse } from '@/types/Pokemon'

export const usePokemonDetails = <SelectedData = PokemonDetailsResponse>(
  name: string,
  select: (data: PokemonDetailsResponse) => SelectedData
) => {
  const { data, isLoading, isError, error } = useQuery<PokemonDetailsResponse, Error, SelectedData>({
    queryKey: ['pokemon', name],
    queryFn: () => fetchPokemonDetails(name),
    select
  })

  return { data: data as SelectedData, isLoading, isError, error }
}
