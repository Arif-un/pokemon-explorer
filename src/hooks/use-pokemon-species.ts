import { useQuery } from '@tanstack/react-query'

import { fetchPokemonSpecies } from '@/services/pokemon-api'

export const usePokemonSpecies = (name: string) => {
  const { data, isLoading, isError, error } = useQuery({
    queryKey: ['pokemon-color', name],
    queryFn: () => fetchPokemonSpecies(name),
    staleTime: 5 * 60 * 1000
  })

  return {
    pokemonColor: data?.color.name,
    isLoading,
    isPokemonSpeciesError: isError,
    pokemonSpeciesError: error
  }
}
