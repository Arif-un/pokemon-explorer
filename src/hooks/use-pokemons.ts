import { useQuery } from '@tanstack/react-query'

import { fetchPokemonWithFilters } from '@/services/pokemon-api-graphql'
import {
  DEFAULT_POKEMON_FETCH_LIMIT,
  DEFAULT_POKEMON_FETCH_OFFSET,
  DEFAULT_QUERY_STALE_TIME
} from '@/utils/constants'

interface UsePokemonsParams {
  name?: string
  limit?: number
  offset?: number
  types?: string
  sort?: string
}

export const usePokemons = ({
  name,
  limit = DEFAULT_POKEMON_FETCH_LIMIT,
  offset = DEFAULT_POKEMON_FETCH_OFFSET,
  types,
  sort
}: UsePokemonsParams) => {
  const { data, isLoading, isError, error, isFetching } = useQuery({
    queryKey: ['pokemon', name, limit, offset, types, sort],
    queryFn: () => fetchPokemonWithFilters({ name, limit, offset, types, sort }),
    staleTime: DEFAULT_QUERY_STALE_TIME,
    placeholderData: previousData => previousData
  })

  return {
    pokemons: data?.pokemons,
    totalItems: data?.pokemon_aggregate.aggregate.count,
    isLoading,
    isError,
    error,
    isFetching
  }
}
