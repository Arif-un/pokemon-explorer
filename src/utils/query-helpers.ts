import { type QueryKey } from '@tanstack/react-query'

import pokemonPlaceholder from '@/assets/Pokeball.svg'
import { type Pokemon, type PokemonDetailsResponse, type PokemonResponse } from '@/types/Pokemon'

import { POKEMON_STATS, TYPE_COLORS, TYPE_ICONS } from './constants'

export interface Stats {
  name: string
  value: number
  max: number
  color: string
}

export interface PokemonTypes {
  name?: string
  icon?: string
  color?: string
}

export function pickPokemonImage(pokemonDetails: Pokemon | PokemonDetailsResponse) {
  const images = Array.isArray(pokemonDetails?.sprites)
    ? pokemonDetails.sprites[0]?.sprites
    : pokemonDetails?.sprites

  return (
    images?.other?.['official-artwork']?.front_shiny ||
    images?.other?.['official-artwork']?.front_default ||
    images?.other?.dream_world?.front_default ||
    images?.other?.home?.front_shiny ||
    images?.front_shiny ||
    pokemonPlaceholder
  )
}

export function preparePokemonDetails(pokemonDetails: PokemonDetailsResponse) {
  const id = pokemonDetails?.id
  const image = pickPokemonImage(pokemonDetails)
  const types = pokemonDetails?.types?.map<PokemonTypes>(typeInfo => ({
    name: typeInfo?.type?.name,
    icon: typeInfo?.type?.name && TYPE_ICONS[typeInfo.type.name],
    color: typeInfo?.type?.name && TYPE_COLORS[typeInfo.type.name]
  }))

  const abilities = pokemonDetails?.abilities?.map(abilityInfo => abilityInfo?.ability?.name)

  const baseStats = pokemonDetails?.stats?.reduce<Stats[]>((acc, statInfo) => {
    if (statInfo?.stat?.name) {
      const statName = statInfo.stat.name
      acc.push({
        name: statName,
        value: statInfo.base_stat || 0,
        max: POKEMON_STATS[statName].max,
        color: POKEMON_STATS[statName].color
      })
    }
    return acc
  }, [])

  return { id, image, types, abilities, baseStats }
}

export function transformTypeFromCacheData(data: { pokemon_v2_type: { name: string } }[]) {
  return data.map(item => ({
    type: {
      name: item.pokemon_v2_type.name
    }
  }))
}

function findPokemonInCache(
  queriesData: [QueryKey, PokemonResponse | undefined][],
  pokemonName: string
) {
  for (const queryData of queriesData) {
    const cachedPokemon = queryData?.[1]?.pokemons?.find(pokemon => pokemon.name === pokemonName)
    if (cachedPokemon) return cachedPokemon
  }
  return null
}

export function preparePokemonDetailsInitialData(
  name: string,
  cachedPokemonQueries: [QueryKey, PokemonResponse | undefined][]
) {
  const cachedPokemon = findPokemonInCache(cachedPokemonQueries, name)

  if (!cachedPokemon) return

  const { sprites, types } = cachedPokemon
  const transformedSprites = sprites?.[0]?.sprites || null
  const transformedTypes = transformTypeFromCacheData(types)

  return {
    sprites: transformedSprites,
    types: transformedTypes
  } as PokemonDetailsResponse
}
