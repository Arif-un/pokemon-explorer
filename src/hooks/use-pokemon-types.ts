import { useQuery } from '@tanstack/react-query'

import { fetchPokemonTypes } from '@/services/pokemon-api-graphql'
import { DEFAULT_QUERY_STALE_TIME } from '@/utils/constants'

export function usePokemonTypes() {
  const { data, isLoading, isError, error } = useQuery({
    queryKey: ['pokemon-types'],
    queryFn: fetchPokemonTypes,
    staleTime: DEFAULT_QUERY_STALE_TIME
  })

  return {
    types: data?.pokemon_v2_type.sort((a, b) => a.name.localeCompare(b.name)),
    isLoading,
    isError,
    error
  }
}
