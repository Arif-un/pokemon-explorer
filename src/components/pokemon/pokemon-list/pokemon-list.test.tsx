import { type HTMLProps } from 'react'

import { render, screen } from '@testing-library/react'
import { describe, expect, it, vi } from 'vitest'

import { useImageLoader } from '@/hooks/use-image-loader'
import { type Pokemon } from '@/types/Pokemon'
import { pickPokemonImage } from '@/utils/query-helpers'

import PokemonList from './pokemon-list'

vi.mock('@tanstack/react-router', () => ({
  Link: ({ children, search, ...props }: HTMLProps<HTMLAnchorElement> & { search: string }) => (
    <a {...props}>{children}</a>
  )
}))

vi.mock('@/hooks/use-image-loader', () => ({
  useImageLoader: vi.fn()
}))

describe('PokemonList', () => {
  const mockPokemons: Pokemon[] = [
    {
      name: 'bulbasaur',
      sprites: [
        {
          sprites: {
            other: {
              'official-artwork': {
                front_shiny: 'https://example.com/pokemon.png'
              }
            }
          }
        }
      ],
      types: [
        {
          pokemon_v2_type: {
            name: 'grass'
          }
        }
      ]
    },
    {
      name: 'charmander',
      sprites: [
        {
          sprites: {
            other: {
              'official-artwork': {
                front_shiny: 'https://example.com/pokemon.png'
              }
            }
          }
        }
      ],
      types: [
        {
          pokemon_v2_type: {
            name: 'grass'
          }
        }
      ]
    },
    {
      name: 'squirtle',
      sprites: [
        {
          sprites: {
            other: {
              'official-artwork': {
                front_shiny: 'https://example.com/pokemon.png'
              }
            }
          }
        }
      ],
      types: [
        {
          pokemon_v2_type: {
            name: 'grass'
          }
        }
      ]
    }
  ]

  it('renders loading state correctly', () => {
    const { asFragment } = render(<PokemonList pokemons={[]} isLoading={true} limit={3} />)
    expect(screen.getAllByTestId('card-skeleton-loader')).toHaveLength(3)
    expect(asFragment()).toMatchSnapshot()
  })

  it('renders pokemon list correctly', () => {
    vi.mocked(useImageLoader).mockReturnValue({
      src: pickPokemonImage(mockPokemons[0]),
      isLoading: false,
      error: null
    })
    const { asFragment } = render(<PokemonList pokemons={mockPokemons} isLoading={false} limit={3} />)

    mockPokemons.forEach(pokemon => {
      expect(screen.getByText(pokemon.name)).toBeInTheDocument()
    })

    expect(asFragment()).toMatchSnapshot()
  })

  it('does not render CardSkeletonLoader when not loading', () => {
    render(<PokemonList pokemons={mockPokemons} isLoading={false} limit={3} />)
    expect(screen.queryByTestId('card-skeleton-loader')).not.toBeInTheDocument()
  })

  it('renders correct number of PokemonPreviewCards', () => {
    render(<PokemonList pokemons={mockPokemons} isLoading={false} limit={3} />)
    mockPokemons.forEach(pokemon => {
      expect(screen.getByLabelText(`View details of ${pokemon.name}`)).toBeInTheDocument()
    })
  })

  it('renders empty list when no pokemons are provided', () => {
    const { container } = render(<PokemonList pokemons={[]} isLoading={false} limit={3} />)
    expect(screen.getByLabelText('List of PokÃ©mon').querySelector('ul')?.childNodes).toHaveLength(0)
    expect(container.firstChild).toMatchSnapshot()
  })

  it('renders PokemonPreviewCards with correct image sources', () => {
    vi.mocked(useImageLoader).mockReturnValue({
      src: pickPokemonImage(mockPokemons[0]),
      isLoading: false,
      error: null
    })
    render(<PokemonList pokemons={mockPokemons} isLoading={false} limit={3} />)
    const images = screen.getAllByRole('img')
    images.forEach(img => {
      expect(img).toHaveAttribute('src', 'https://example.com/pokemon.png')
    })
  })
})
