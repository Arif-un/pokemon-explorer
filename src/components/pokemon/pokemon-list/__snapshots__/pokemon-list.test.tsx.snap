// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PokemonList > renders empty list when no pokemons are provided 1`] = `
<section
  aria-label="List of Pokémon"
  class="mt-56 p-2"
>
  <ul
    class="flex justify-center gap-x-10 gap-y-28 flex-wrap"
  />
</section>
`;

exports[`PokemonList > renders loading state correctly 1`] = `
<DocumentFragment>
  <section
    aria-label="List of Pokémon"
    class="mt-56 p-2"
  >
    <ul
      class="flex justify-center gap-x-10 gap-y-28 flex-wrap"
    >
      <div
        class="w-60 h-[258px] flex justify-center items-center shadow border border-slate-300 text-slate-400 bg-slate-300 dark:bg-slate-800 dark:text-slate-950 dark:border-slate-700 rounded-3xl animate-pulse "
        data-testid="card-skeleton-loader"
      >
        <svg
          fill="currentColor"
          height="800px"
          viewBox="0 0 512 512"
          width="800px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M450.46,256.09C449.35,175.17,399.81,102.71,324,73.79,247.59,44.67,157.49,69,105.82,132.13,54.4,195,46.61,285.58,88.49,355.68c41.8,69.95,123.74,106,203.55,91.63,91-16.37,156.14-98.12,158.35-189.14A20.16,20.16,0,0,0,450.46,256.09ZM119.05,174.38C152.76,118,220.23,87,285,99.43c69.4,13.29,120.43,70.47,128.83,139H318.41c-8.26-27.36-32-48-62.62-48-29.65,0-55.15,20.65-63.11,48H97.74A158,158,0,0,1,119.05,174.38ZM286.13,256.1c-2,38.75-60.67,39.4-60.67,0S284.17,217.33,286.13,256.1Zm24,149.79C246.85,428.58,175,408.74,132.3,356.82a157.53,157.53,0,0,1-34.57-83H192.6c7.91,27.39,33.7,48,63.19,48,30.67,0,54.36-20.68,62.62-48h95.45C406.61,333,367.54,385.32,310.14,405.89Z"
          />
        </svg>
      </div>
      <div
        class="w-60 h-[258px] flex justify-center items-center shadow border border-slate-300 text-slate-400 bg-slate-300 dark:bg-slate-800 dark:text-slate-950 dark:border-slate-700 rounded-3xl animate-pulse "
        data-testid="card-skeleton-loader"
      >
        <svg
          fill="currentColor"
          height="800px"
          viewBox="0 0 512 512"
          width="800px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M450.46,256.09C449.35,175.17,399.81,102.71,324,73.79,247.59,44.67,157.49,69,105.82,132.13,54.4,195,46.61,285.58,88.49,355.68c41.8,69.95,123.74,106,203.55,91.63,91-16.37,156.14-98.12,158.35-189.14A20.16,20.16,0,0,0,450.46,256.09ZM119.05,174.38C152.76,118,220.23,87,285,99.43c69.4,13.29,120.43,70.47,128.83,139H318.41c-8.26-27.36-32-48-62.62-48-29.65,0-55.15,20.65-63.11,48H97.74A158,158,0,0,1,119.05,174.38ZM286.13,256.1c-2,38.75-60.67,39.4-60.67,0S284.17,217.33,286.13,256.1Zm24,149.79C246.85,428.58,175,408.74,132.3,356.82a157.53,157.53,0,0,1-34.57-83H192.6c7.91,27.39,33.7,48,63.19,48,30.67,0,54.36-20.68,62.62-48h95.45C406.61,333,367.54,385.32,310.14,405.89Z"
          />
        </svg>
      </div>
      <div
        class="w-60 h-[258px] flex justify-center items-center shadow border border-slate-300 text-slate-400 bg-slate-300 dark:bg-slate-800 dark:text-slate-950 dark:border-slate-700 rounded-3xl animate-pulse "
        data-testid="card-skeleton-loader"
      >
        <svg
          fill="currentColor"
          height="800px"
          viewBox="0 0 512 512"
          width="800px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M450.46,256.09C449.35,175.17,399.81,102.71,324,73.79,247.59,44.67,157.49,69,105.82,132.13,54.4,195,46.61,285.58,88.49,355.68c41.8,69.95,123.74,106,203.55,91.63,91-16.37,156.14-98.12,158.35-189.14A20.16,20.16,0,0,0,450.46,256.09ZM119.05,174.38C152.76,118,220.23,87,285,99.43c69.4,13.29,120.43,70.47,128.83,139H318.41c-8.26-27.36-32-48-62.62-48-29.65,0-55.15,20.65-63.11,48H97.74A158,158,0,0,1,119.05,174.38ZM286.13,256.1c-2,38.75-60.67,39.4-60.67,0S284.17,217.33,286.13,256.1Zm24,149.79C246.85,428.58,175,408.74,132.3,356.82a157.53,157.53,0,0,1-34.57-83H192.6c7.91,27.39,33.7,48,63.19,48,30.67,0,54.36-20.68,62.62-48h95.45C406.61,333,367.54,385.32,310.14,405.89Z"
          />
        </svg>
      </div>
    </ul>
  </section>
</DocumentFragment>
`;

exports[`PokemonList > renders pokemon list correctly 1`] = `
<DocumentFragment>
  <section
    aria-label="List of Pokémon"
    class="mt-56 p-2"
  >
    <ul
      class="flex justify-center gap-x-10 gap-y-28 flex-wrap"
    >
      <li
        class="list-none"
      >
        <a
          aria-label="View details of bulbasaur"
          class="inline-block focus-visible:outline outline-yellow-500 outline-2 focus-visible:outline-offset-8 outline-offset-0  rounded-3xl transition-offset duration-200 motion-reduce:duration-0"
          to="pokemon-details/bulbasaur"
        >
          <div
            class="inner-glow bg-slate-800 inline-flex flex-col justify-center group cursor-pointer  dark:bg-slate-900 p-2 dark:text-slate-50 rounded-3xl dark:border-slate-800 border"
          >
            <div
              class="parallax-effect-glare-scale "
            >
              <figure
                class="popup-element flex items-center justify-center"
              >
                <div
                  class="relative w-60 h-60 -mt-16"
                >
                  <img
                    alt="bulbasaur image-backdrop"
                    class="blur-2xl w-52 absolute opacity-50 group-hover:opacity-90 m-auto inset-0 transition-opacity"
                    height="240"
                    loading="lazy"
                    src="https://example.com/pokemon.png"
                    width="240"
                  />
                  <img
                    alt="bulbasaur image"
                    class="w-52 m-auto z-10 relative"
                    height="240"
                    loading="lazy"
                    src="https://example.com/pokemon.png"
                    width="240"
                  />
                </div>
              </figure>
              <div
                class="flex items-center justify-center gap-2"
              >
                <h3
                  class="text-slate-300 group-hover:text-slate-50 dark:text-slate-400 dark:group-hover:text-slate-50 group-hover:underline capitalize text-center font-semibold my-5 text-xl"
                >
                  bulbasaur
                </h3>
                <svg
                  aria-hidden="true"
                  aria-label="navigate icon"
                  class="lucide lucide-arrow-up-right opacity-0 size-0 scale-0 group-hover:opacity-100 group-hover:scale-100 group-hover:size-auto transition-transform text-slate-50 dark:text-slate-50"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M7 7h10v10"
                  />
                  <path
                    d="M7 17 17 7"
                  />
                </svg>
              </div>
              <div
                class="glare-wrapper"
                style="position: absolute; top: 0px; left: 0px; width: 100%; height: 100%; overflow: hidden; border-radius: 24px; pointer-events: none;"
              >
                <div
                  class="glare"
                  style="position: absolute; top: 50%; left: 50%; transform-origin: 0% 0%; pointer-events: none; width: 0px; height: 0px;"
                />
              </div>
            </div>
          </div>
        </a>
      </li>
      <li
        class="list-none"
      >
        <a
          aria-label="View details of charmander"
          class="inline-block focus-visible:outline outline-yellow-500 outline-2 focus-visible:outline-offset-8 outline-offset-0  rounded-3xl transition-offset duration-200 motion-reduce:duration-0"
          to="pokemon-details/charmander"
        >
          <div
            class="inner-glow bg-slate-800 inline-flex flex-col justify-center group cursor-pointer  dark:bg-slate-900 p-2 dark:text-slate-50 rounded-3xl dark:border-slate-800 border"
          >
            <div
              class="parallax-effect-glare-scale "
            >
              <figure
                class="popup-element flex items-center justify-center"
              >
                <div
                  class="relative w-60 h-60 -mt-16"
                >
                  <img
                    alt="charmander image-backdrop"
                    class="blur-2xl w-52 absolute opacity-50 group-hover:opacity-90 m-auto inset-0 transition-opacity"
                    height="240"
                    loading="lazy"
                    src="https://example.com/pokemon.png"
                    width="240"
                  />
                  <img
                    alt="charmander image"
                    class="w-52 m-auto z-10 relative"
                    height="240"
                    loading="lazy"
                    src="https://example.com/pokemon.png"
                    width="240"
                  />
                </div>
              </figure>
              <div
                class="flex items-center justify-center gap-2"
              >
                <h3
                  class="text-slate-300 group-hover:text-slate-50 dark:text-slate-400 dark:group-hover:text-slate-50 group-hover:underline capitalize text-center font-semibold my-5 text-xl"
                >
                  charmander
                </h3>
                <svg
                  aria-hidden="true"
                  aria-label="navigate icon"
                  class="lucide lucide-arrow-up-right opacity-0 size-0 scale-0 group-hover:opacity-100 group-hover:scale-100 group-hover:size-auto transition-transform text-slate-50 dark:text-slate-50"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M7 7h10v10"
                  />
                  <path
                    d="M7 17 17 7"
                  />
                </svg>
              </div>
              <div
                class="glare-wrapper"
                style="position: absolute; top: 0px; left: 0px; width: 100%; height: 100%; overflow: hidden; border-radius: 24px; pointer-events: none;"
              >
                <div
                  class="glare"
                  style="position: absolute; top: 50%; left: 50%; transform-origin: 0% 0%; pointer-events: none; width: 0px; height: 0px;"
                />
              </div>
            </div>
          </div>
        </a>
      </li>
      <li
        class="list-none"
      >
        <a
          aria-label="View details of squirtle"
          class="inline-block focus-visible:outline outline-yellow-500 outline-2 focus-visible:outline-offset-8 outline-offset-0  rounded-3xl transition-offset duration-200 motion-reduce:duration-0"
          to="pokemon-details/squirtle"
        >
          <div
            class="inner-glow bg-slate-800 inline-flex flex-col justify-center group cursor-pointer  dark:bg-slate-900 p-2 dark:text-slate-50 rounded-3xl dark:border-slate-800 border"
          >
            <div
              class="parallax-effect-glare-scale "
            >
              <figure
                class="popup-element flex items-center justify-center"
              >
                <div
                  class="relative w-60 h-60 -mt-16"
                >
                  <img
                    alt="squirtle image-backdrop"
                    class="blur-2xl w-52 absolute opacity-50 group-hover:opacity-90 m-auto inset-0 transition-opacity"
                    height="240"
                    loading="lazy"
                    src="https://example.com/pokemon.png"
                    width="240"
                  />
                  <img
                    alt="squirtle image"
                    class="w-52 m-auto z-10 relative"
                    height="240"
                    loading="lazy"
                    src="https://example.com/pokemon.png"
                    width="240"
                  />
                </div>
              </figure>
              <div
                class="flex items-center justify-center gap-2"
              >
                <h3
                  class="text-slate-300 group-hover:text-slate-50 dark:text-slate-400 dark:group-hover:text-slate-50 group-hover:underline capitalize text-center font-semibold my-5 text-xl"
                >
                  squirtle
                </h3>
                <svg
                  aria-hidden="true"
                  aria-label="navigate icon"
                  class="lucide lucide-arrow-up-right opacity-0 size-0 scale-0 group-hover:opacity-100 group-hover:scale-100 group-hover:size-auto transition-transform text-slate-50 dark:text-slate-50"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M7 7h10v10"
                  />
                  <path
                    d="M7 17 17 7"
                  />
                </svg>
              </div>
              <div
                class="glare-wrapper"
                style="position: absolute; top: 0px; left: 0px; width: 100%; height: 100%; overflow: hidden; border-radius: 24px; pointer-events: none;"
              >
                <div
                  class="glare"
                  style="position: absolute; top: 50%; left: 50%; transform-origin: 0% 0%; pointer-events: none; width: 0px; height: 0px;"
                />
              </div>
            </div>
          </div>
        </a>
      </li>
    </ul>
  </section>
</DocumentFragment>
`;
